File: app/src/main/java/com/example/buildmypc/ui/parts/parts/Cooler.java
package com.example.buildmypc.ui.parts.parts;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;

import static android.os.Build.VERSION_CODES.Q;

public final class Cooler extends Part implements Parcelable {
//	private final String manufacturer;
//	private final String model;
	private final String rpm;
	private final String noiseLevel;
	private final int height;
	private final ArrayList<String> socketSupport;
	private final boolean waterCooled;
	private final boolean fanless;

	public Cooler(String model, String manufacturer) {
		super(model, manufacturer);
		this.rpm = null;
		this.noiseLevel = null;
		this.height = -1;
		this.socketSupport = null;
		this.waterCooled = false;
		this.fanless = false;
	}

	public Cooler(String model, String manufacturer, String rpm, String noiseLevel, int height, ArrayList<String> socketSupport, boolean waterCooled, boolean fanless) {
		super(model, manufacturer);
		this.rpm = rpm;
		this.noiseLevel = noiseLevel;
		this.height = height;
		this.socketSupport = socketSupport;
		this.waterCooled = waterCooled;
		this.fanless = fanless;
	}

	@RequiresApi(Q)
	public Cooler(@NotNull Parcel in) {
		super(in.readString(), in.readString());
		rpm = in.readString();
		noiseLevel = in.readString();
		height = in.readInt();
		socketSupport = in.createStringArrayList();
		waterCooled = in.readBoolean();
		fanless = in.readBoolean();
	}

	public String getRpm() {
		return rpm;
	}

	public String getNoiseLevel() {
		return noiseLevel;
	}

	public int getHeight() {
		return height;
	}

	public ArrayList<String> getSocketSupport() {
		return socketSupport;
	}

	public boolean isWaterCooled() {
		return waterCooled;
	}

	public boolean isFanless() {
		return fanless;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;
		if (!super.equals(o)) return false;

		Cooler cooler = (Cooler) o;

		if (getHeight() != cooler.getHeight()) return false;
		if (isWaterCooled() != cooler.isWaterCooled()) return false;
		if (isFanless() != cooler.isFanless()) return false;
		if (getRpm() != null ? !getRpm().equals(cooler.getRpm()) : cooler.getRpm() != null)
			return false;
		if (getNoiseLevel() != null ? !getNoiseLevel().equals(cooler.getNoiseLevel()) : cooler.getNoiseLevel() != null)
			return false;
		return getSocketSupport() != null ? getSocketSupport().equals(cooler.getSocketSupport()) : cooler.getSocketSupport() == null;
	}

	@Override
	public int hashCode() {
		int result = super.hashCode();
		result = 31 * result + (getRpm() != null ? getRpm().hashCode() : 0);
		result = 31 * result + (getNoiseLevel() != null ? getNoiseLevel().hashCode() : 0);
		result = 31 * result + getHeight();
		result = 31 * result + (getSocketSupport() != null ? getSocketSupport().hashCode() : 0);
		result = 31 * result + (isWaterCooled() ? 1 : 0);
		result = 31 * result + (isFanless() ? 1 : 0);
		return result;
	}

	@RequiresApi(Q)
	@Override
	public void writeToParcel(@NotNull Parcel dest, int flags) {
		super.writeToParcel(dest, flags);
		dest.writeString(rpm);
		dest.writeString(noiseLevel);
		dest.writeInt(height);
		dest.writeStringList(socketSupport);
		dest.writeBoolean(waterCooled);
		dest.writeBoolean(fanless);
	}

	@NonNull
	@NotNull
	@Override
	public String toString() {
		return "Cooler " + getModel() + " " + getManufacturer();
	}
}

File: app/src/main/java/com/example/buildmypc/ui/parts/parts/Cooler.kt
package com.example.buildmypc.ui.parts.parts

import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator
import org.jetbrains.annotations.Contract
import java.util.*

class Cooler : Part, Parcelable {
	val rpm: String?
	val noiseLevel: String?
	val height: Int
	val socketSupport: ArrayList<String>?
	val isWaterCooled: Boolean
	val isFanless: Boolean

	constructor(model: String?, manufacturer: String?) : super(model, manufacturer) {
		this.rpm = null
		this.noiseLevel = null
		this.height = -1
		this.socketSupport = null
		this.isWaterCooled = false
		this.isFanless = false
	}

	constructor(
		model: String?,
		manufacturer: String?,
		rpm: String?,
		noiseLevel: String?,
		height: Int,
		socketSupport: ArrayList<String>?,
		waterCooled: Boolean,
		fanless: Boolean
	) : super(model, manufacturer) {
		this.rpm = rpm
		this.noiseLevel = noiseLevel
		this.height = height
		this.socketSupport = socketSupport
		this.isWaterCooled = waterCooled
		this.isFanless = fanless
	}

	constructor(`in`: Parcel) : super(`in`.readString(), `in`.readString()) {
		this.rpm = `in`.readString()
		this.noiseLevel = `in`.readString()
		this.height = `in`.readInt()
		this.socketSupport = `in`.createStringArrayList()
		this.isWaterCooled = `in`.readBoolean()
		this.isFanless = `in`.readBoolean()
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other == null || javaClass != other.javaClass) return false
		if (!super.equals(other)) return false
		val cooler = other as Cooler
		if (height != cooler.height) return false
		if (isWaterCooled != cooler.isWaterCooled) return false
		if (isFanless != cooler.isFanless) return false
		if (if (rpm != null) rpm != cooler.rpm else cooler.rpm != null) return false
		if (if (noiseLevel != null) noiseLevel != cooler.noiseLevel else cooler.noiseLevel != null) return false
		return if (socketSupport != null) socketSupport == cooler.socketSupport else cooler.socketSupport == null
	}

	override fun hashCode(): Int {
		var result = super.hashCode()
		result = 31 * result + (rpm?.hashCode() ?: 0)
		result = 31 * result + (noiseLevel?.hashCode() ?: 0)
		result = 31 * result + height
		result = 31 * result + (socketSupport?.hashCode() ?: 0)
		result = 31 * result + if (isWaterCooled) 1 else 0
		result = 31 * result + if (isFanless) 1 else 0
		return result
	}

	override fun writeToParcel(dest: Parcel, flags: Int) {
		super.writeToParcel(dest, flags)
		dest.writeString(rpm)
		dest.writeString(noiseLevel)
		dest.writeInt(height)
		dest.writeStringList(socketSupport)
		dest.writeBoolean(isWaterCooled)
		dest.writeBoolean(isFanless)
	}

	override val paramCount = 8

	companion object {
		@JvmField
		val CREATOR: Creator<Cooler> = object : Creator<Cooler> {
			@Contract("_ -> new")
			override fun createFromParcel(`in`: Parcel) = Cooler(`in`)

			@Contract(value = "_ -> new", pure = true)
			override fun newArray(size: Int): Array<Cooler?> = arrayOfNulls(size)
		}
	}
}